### PR TITLE
 test: add shape validation and chain ID tests for pending transactions

### PR DESCRIPTION
**Pull request by rabi-siddique**
## Description

**Testing Improvements:**
• Added tests for `processPendingTxEvents` validating `PublishedTxShape` across all transaction types (`CCTP_TO_EVM`, `CCTP_TO_NOBLE`, `GMP`). Also removing `published-tx-shape.test.ts` because it wasn’t validating the functions used in the off-chain resolver.
• Tests required vs optional fields - `amount` required for CCTP, optional for GMP
• Covers edge cases: missing fields, wrong types, invalid enums, negative amounts
• Added `nobleWithdrawMonitor` chain validation tests for allowed Noble chains (`noble-1`, `grand-1`)

**Bug Fixes:**
• Replaced `toSorted()` with `sort()` for broader Node.js compatibility
• Added `nobleChainIds` constant for consistent chain validation instead of hardcoded references.

**Refactoring:**
• Extracted common fields in `PublishedTxShape` to reduce duplication
• Made `createMockPendingTxData` flexible - made `type` required, removed `amount` defaulting
• Fixed mock utility to only include `amount` when explicitly provided

### Testing Considerations

Unit tests have been added for the new validation logic and for validating PendingTx shapes.
 
### Upgrade Considerations

This PR includes a fix that is required for mainnet use:
- The `nobleChainIds` bug fix prevents resolver failures during Noble transfer watching - previously invalid chain references would cause the resolver to fail at runtime.
- Node.js compatibility fix supports both old and new Node.js versions

### CODE REVIEW

File: services/ymax-planner/src/resolver.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ const getInvitationMakers = async (wallet: SigningSmartWalletKit) => {
   const getCurrentWalletRecord = await wallet.query.getCurrentWalletRecord();
   const invitation = getCurrentWalletRecord.offerToUsedInvitation
     .filter(inv => inv[1].value[0].description === 'resolver')
-    .toSorted()
+    .sort()
```
Comment (dckc) [inline services/ymax-planner/src/resolver.ts:17]: interesting... a search for `toSorted` in `agoric-sdk/packages/xsnap` doesn't turn up anything.

the version we're using seems to be:

2023-05-30 14:08 [f6c5951fc](https://github.com/Moddable-OpenSource/moddable/tree/f6c5951fc055e4ca592b9166b9ae3cbb9cca6bf0)

Node.js compatibility seems to be a red herring; `toSorted()` is available in node 20, which is the minimum version in [agoric-sdk prerequisites](https://github.com/Agoric/agoric-sdk#prerequisites).

File: services/ymax-planner/src/pending-tx-manager.ts [hunk 2]
```diff
@@ -137,9 +137,24 @@ const nobleWithdrawMonitor: PendingTxMonitor<NobleWithdrawTx, EvmContext> = {
...
+     * @description
+     * A list of Noble chain IDs used on both mainnet and testnet networks.
+     *
+     * **Mainnet Chain**
+     * - `noble-1` — Status: https://noble-rpc.polkachu.com/status
+     *
+     * **Testnet Chains**
+     * - `grand-1` — Status: https://rpc.testnet.noble.xyz/status
+     */
+    const nobleChainIds = ['noble-1', 'grand-1'];
```
Comment (dckc) [inline services/ymax-planner/src/pending-tx-manager.ts:153]: I suggest we should query `ctx.cosmosRest` for the noble chainId at run-time.
Optimally, do it when the service starts.

File: services/ymax-planner/test/pending-tx-shape.test.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+const mockEvmCtx = createMockEvmContext();
+const powers = {
+  ...mockEvmCtx,
+  marshaller,
+  log: () => {},
+  error: () => {},
+};
+const mockHandlePendingTxFn = async () => {};
+
+test('processPendingTxEvents - valid CCTP_TO_EVM shape passes mustMatch', async t => {
```
Comment (dckc) [inline services/ymax-planner/test/pending-tx-shape.test.ts:23]: if the test is checking a shape, why call `processPendingTxEvents`? why not just call `mustMatch` in the test?

by way of example, see https://github.com/Agoric/agoric-sdk/blob/master/packages/portfolio-contract/test/offer-shapes.test.ts

File: services/ymax-planner/test/pending-tx-shape.test.ts [hunk 4]
```diff
@@ -101,41 +101,3 @@ test('processPendingTxEvents - missing required amount field for CCTP_TO_NOBLE f
     errorMessages.some(msg => msg.includes('Failed to process pending tx')),
   );
 });
-
-test('processPendingTxEvents - invalid tx type fails mustMatch', async t => {
```
Comment (dckc) [inline services/ymax-planner/test/pending-tx-shape.test.ts]: was this test added in the same PR?

File: packages/portfolio-contract/src/resolver/types.ts [hunk 5]
```diff
@@ -48,36 +48,28 @@ export type PublishedTx = {
...
 };
 
-/**
- * A PendingTx is a PublishedTx (published by ymax contract) with an additional
- * txId property used by the resolver to track and manage pending transactions.
- */
-export type PendingTx = { txId: TxId } & PublishedTx;
+const publishedTxRequiredFields = harden({
+  // Format: `${chainId}:${chainId}:${remoteAddess}`
+  destinationAddress: M.string(),
```
Comment (dckc) [inline packages/portfolio-contract/src/resolver/types.ts:59]: **suggestion**: call it `destinationAccountId`; that makes the comment about format unnecessary

File: packages/portfolio-contract/src/resolver/types.ts [hunk 6]
```diff
@@ -48,36 +48,28 @@ export type PublishedTx = {
...
+  status: M.or(TxStatus.PENDING),
+});
 
+// `amount` is optional for GMP but required for CCTP
+const publishedTxCctpRequiredFields = { amount: M.nat() };
 export const PublishedTxShape: TypedPattern<PublishedTx> = M.or(
```
Comment (dckc) [inline packages/portfolio-contract/src/resolver/types.ts:65]: This shape is now different from the type.

I'm upgrading my [Aug 28 suggestion](https://github.com/Agoric/agoric-sdk/pull/11818#discussion_r2308796793) to a change request: please use a discriminated union for `PublishedTx`.

I think
 - every PublishedTx has a status, destinationAddress, and type
 - depending on the type, it may have additional fields. GMP has none. CCTP has amount.

In typescript, that's:

```ts
type CCTPType = TxType.CCTP_TO_EVM | TxType.CCTP_TO_NOBLE;
export type PublishedTx = { status: TxStatus; destinationAddress: AccountId } & (
  | { type: CCTP_TYPE, amount: NatValue }
  | { type: TxType.GMP }
);
```

As an Endo pattern, thats:

```ts
const CCTPTypeShape: TypedPattern<CCTPType> = M.or(TxType.CCTP_TO_EVM, TxType.CCTP_TO_NOBLE);

const PublishedTxShape: TypedPattern<PublishedTx> = M.splitRecord({
  status: M.or(TxStatus.PENDING),
  destinationAddress: M.string(), // XXX AccountId format not validated
}, {}, M.or({ type: CCTPTypeShape, amount: M.nat() }, { type: TxType.GMP } ))
```

see also:

 - #6115

File: packages/portfolio-contract/tools/mocks.ts [hunk 7]
```diff
@@ -1,22 +1,24 @@
...
 
-import { TxType, type TxStatus } from '../src/resolver/constants.js';
+type MockPendingTx = {
+  type: string;
+  amount?: bigint;
+  status?: string;
+  destinationAddress?: AccountId;
+};
 
 export const createMockPendingTxData = ({
```
Comment (dckc) [inline packages/portfolio-contract/tools/mocks.ts:11]: I don't understand the purpose of `MockPendingTx` and `createMockPendingTxData`. Just writing constant expressions in the tests seems like it would be easier to read.

**Review by dckc — CHANGES_REQUESTED** [review]
The change to `PublishedTxShape` makes it clear that it's different from the `PublishedTx` type. Please use a discriminated union in both cases.

**Comment by rabi-siddique** [comment]
Not relevant anymore. Updates introduced in this PR were applied through other PRs.
